photosynthesis_metrics/vif.py

```diff
@@ -0,0 +1,138 @@
+r"""Implemetation of Visual Information Fidelity metric
+Code is based on MATLAB version for computations in pixel domain
+https://live.ece.utexas.edu/research/Quality/VIF.htm
+
+References:
+    https://ieeexplore.ieee.org/abstract/document/1576816/
+"""
+import torch
+import numpy as np
+from torch.nn.modules.loss import _Loss
+import torch.nn.functional as F
+
+from photosynthesis_metrics.utils import _adjust_dimensions, _validate_input
+
+
+def _gaussian_kernel2d(kernel_size: int = 5, sigma: float = 2.0) -> torch.Tensor:
+    r"""Returns 2D Gaussian kernel N(0,`sigma`)
+    Args:
+        kernel_size: Size
+        sigma: Sigma
+    Returns:
+        gaussian_kernel: 2D kernel with shape (kernel_size x kernel_size)
+        
+    """
+    x = torch.arange(- (kernel_size // 2), kernel_size // 2 + 1).resize(1, kernel_size)
+    y = torch.arange(- (kernel_size // 2), kernel_size // 2 + 1).resize(kernel_size, 1)
+    kernel = torch.exp(-(x * x + y * y) / (2.0 * sigma ** 2))
+    # Normalize
+    kernel = kernel / torch.sum(kernel)
+    return kernel
+
+
+def vif_p(prediction: torch.Tensor, target: torch.Tensor,
+          sigma_n_sq: float = 2.0, data_range: int = 1.0) -> torch.Tensor:
+    r"""Compute Visiual Information Fidelity in **pixel** domain for a batch of images.
+    This metric isn't symmetric, so make sure to place arguments in correct order.
+
+    Both inputs supposed to be in range [0, 1] with RGB order.
+    Args:
+        prediction: Batch of predicted images with shape (batch_size x channels x H x W)
+        target: Batch of target images with shape  (batch_size x channels x H x W)
+        sigma_n_sq: HVS model parameter (variance of the visual noise).
+        data_range: Value range of input images (usually 1.0 or 255). Default: 1.0
+        
+    Returns:
+        VIF: Index of similarity betwen two images. Usually in [0, 1] interval.
+            Can be bigger than 1 for predicted images with higher contrast than original one.
+    Note:
+        In original paper this method was used for bands in discrete wavelet decomposition.
+        Later on authors released code to compute VIF approximation in pixel domain.
+        See https://live.ece.utexas.edu/research/Quality/VIF.htm for details.
+        
+    """
+    if data_range == 255:
```

Comment:
Current `_validate_input` allows 5D tensor of `(N,C,H,W,2)` size. 
```suggestion
    assert prediction.dim() == 4, f'Expected 4D tensor, got {prediction.size()}'
    if data_range == 255:
```